mm/context_label.py

```diff
@@ -0,0 +1,161 @@
+
+import numpy as np
+
+from i6_core.lib.hdf import get_returnn_simple_hdf_writer
+from i6_core.lib.rasr_cache import FileArchive
+from sisyphus import Job, Task, tk, gs
+from apptek_asr.artefacts.factory import AbstractArtefactRepository
+
+
+class GetContextLabelFromDenseTyingJob(Job):
+    def __init__(
+        self,
+        alignment_cache_path: tk.Path,
+        allophone_path: tk.Path,
+        dense_tying_path: tk.Path,
+        n_contexts: int,
+        returnn_root: Optional[tk.Path] = None,
+    ):
+        """
+        Get past/center/future context label of alignment by calculating back labels from
+        dense tying and write the labels into hdf file.
+        (C.f. NoStateTyingDense in rasr
+        https://github.com/rwth-i6/rasr/blob/a942e3940c30eeba900c873f3bfb3f48d5b39ddb/src/Am/ClassicStateTying.cc#L272)
+
+
+        :param alignment_cache_path: path to alginment cache
+        :param allophone_path: path to allohone
+        :param dense_tying_path: path to denser tying file (hint: can be obatained by setting
+            crp.acoustic_model_config.state_tying.type = "no-tying-dense"
+            crp.acoustic_model_config.state_tying.use_boundary_classes = "no"
+            crp.acoustic_model_config.state_tying.use_word_end_classes = "no")
+        :param n_contexts: number of phonemes in lexicon + 1 (for non-context # in rasr)
+        :param returnn_root: path to returnn root
+        """
+        self.alignment_cache_path = alignment_cache_path
+        self.allophone_path = allophone_path
+        self.dense_tying_path = dense_tying_path
+        self.n_contexts = n_contexts
+        self.returnn_root = returnn_root
+
+        self.out_hdf_left_context = self.output_path("left_context.hdf")
+        self.out_hdf_right_context = self.output_path("right_context.hdf")
+        self.out_hdf_center_context = self.output_path("center_context.hdf")
+
+        self.rqmt = {"cpu": 1, "mem": 8, "time": 0.5}
+
+    def tasks(self):
+        yield Task("run", resume="run", rqmt=self.rqmt)
+
+    @classmethod
+    def get_tying_and_num_classes(cls, dense_tying_path: tk.Path) -> Tuple[Dict, int]:
+        num_classes = 0
+        for line in open(dense_tying_path.get_path(), "rt"):
+            if not line.startswith("#"):
+                num_classes = max(num_classes, int(line.strip().split()[1]))
+
+        state_tying = dict((k, int(v)) for l in open(dense_tying_path.get_path()) for k, v in [l.strip().split()[0:2]])
+
+        return state_tying, num_classes
+
+    @classmethod
+    def get_target_labels_from_dense(cls, dense_label: int, hmm_state: int, n_contexts: int) -> Tuple[int, int, int]:
+        """ """
+
+        futureLabel = np.mod(dense_label, n_contexts)
+        popFutureLabel = np.floor_divide(dense_label, n_contexts)
+
+        pastLabel = np.mod(popFutureLabel, n_contexts)
+        centerState = np.floor_divide(popFutureLabel, n_contexts)
+        centerState = np.floor_divide(centerState - hmm_state, 3)
```

Comment:
This is not correct. Please see also the comment above.

Once the identity of right and left are popped, one is left with the center phoneme state information. Depending on how you dumped the `state-tying` options in `RASR`, you would do further steps to pop also word end class and hmm state info.
My suggestion here is that the class has a dictionary that has the label info, such as whether end of word or boundary classes are used, and what is the number of hmm state per phoneme. And then to apply the exact inverse function with respect to the implementation of RASR [here](https://github.com/rwth-i6/rasr/blob/master/src/Am/ClassicStateTying.cc#L272)

Reply:
or just parse the string representation of the allophone?